### PR TITLE
fixes #4806 add support to register compute resource provider from a plugin

### PR DESCRIPTION
**Pull request by abenari**

### CODE REVIEW

File: app/models/compute_resource.rb [hunk 1]
```diff
@@ -37,10 +39,19 @@ class ComputeResource < ActiveRecord::Base
...
+  def self.providers
+    supported_providers.reject { |p| !SETTINGS[p.downcase.to_sym] }
+  end
+
   # allows to create a specific compute class based on the provider.
   def self.new_provider args
-    raise ::Foreman::Exception.new(N_("must provide a provider")) unless provider = args[:provider]
-    PROVIDERS.each do |p|
+    raise ::Foreman::Exception.new(N_("must provide a provider")) unless provider = args.delete(:provider)
+    self.class.providers.each do |p|
```
Comment (domcleal) [inline app/models/compute_resource.rb]: remove "self.class" I think

**Comment by domcleal** [comment]
:+1: this will be very useful.  Do you have an example?  Are any other workarounds needed?

The only additional thing I'd like to see are some tests around the ComputeResource class/model methods and one for registration in test/unit/plugin_test.rb.

**Comment by abenari** [comment]
A working example is:
https://github.com/abenari/foreman-docker
I am working on tests ..

File: app/services/foreman/plugin.rb [hunk 2]
```diff
@@ -186,5 +186,11 @@ def role(name, permissions)
       end
     end
 
+    # Add Compute resource
+    def compute_resource name
```
Comment (isratrade) [inline app/services/foreman/plugin.rb]: nitpik `def compute_resource(name)` seems more readable

**Comment by isratrade** [comment]
@abenari, looks good

**Comment by abenari** [comment]
Now include tests.

**Comment by ohadlevy** [comment]
I would like to extract the sti prefix, so plugins are not bounded to the foreman/model module path.

**Comment by domcleal** [comment]
@ohadlevy is Provider::Compute what you meant?  I took your comment to mean we should support providers in any module namespace, not just Foreman::Model.

**Comment by ohadlevy** [comment]
I had a discussion about this with @abenari, supporting any prefix might be
more complicated, for example:

if we have auth code that checks for all CR classes, its easier to check
the filed name to start with STI_PREFIX rather then try to constantize them:

I was mostly trying to avoid people to have a directory structure which is:
app/model/foreman/model/new_cr.rb

this way, we will keep a common STI prefix (needed to avoid conflicts with
existing ruby classes/modules - e.g. docker, libvirt etc) and wont force
the directory structure.

the reason for Provider::Compute is that we might have other provides
(storage, network etc).

does that make sense? any idea for improvements?

**Comment by domcleal** [comment]
I think we could simply have plugins register the full class name of their compute resource.  The use of STI_PREFIX looks like a shortcut so we only have to store the short name ("Libvirt") in the list of `supported_providers` and as returned from `providers`.

If we instead registered CRs with both a name and a class, then store the name + class.name in the supported_providers array, then I think we can remove STI_PREFIX.  Or just register with the class, and take `class.name.split('::').last` as the name of the provider.

**Comment by ohadlevy** [comment]
the original issue, is that ActiveRecord tries to find the class name based
on the type column.

storing a type Libvirt would cause a conflict, as it would find the Libvirt
module (ruby-libvirt gem).

**Comment by domcleal** [comment]
I thought the type column stored the fully qualified class name, not the single class name.

**Comment by ohadlevy** [comment]
thats right.

**Comment by domcleal** [comment]
Right, so I wasn't proposing changing that at all.. the only use of STI_PREFIX appears to be linking between our list of known providers (which are simple names, e.g. "Libvirt") from the UI or wherever to classes.  Instead of assuming STI_PREFIX, when plugins register, they could register the class name.

I was thinking of this: https://github.com/domcleal/foreman/commit/6dd0755ceed2c36eaeffd2b1e0d9ff15928cad95

(though I seem to have broken auditing, as `Foreman::Model::EC2.reflect_on_association(:audits)` now returns nil for reasons I can't figure out.)

**Comment by abenari** [comment]
We do assume a directory structure for compute resource ui files, adding directory structure to the one model class per cr, will make it easy to find.

**Comment by domcleal** [comment]
I'm happy with a convention (though I think taking a top-level namespace like "Provider" is bad, Foreman was better..), but not sure it needs to be forced?

**Comment by domcleal** [comment]
Whichever way, this works great :+1:

I look forward to losing some fat in Foreman core and getting some new things (maybe CloudStack or Xen?) into plugins to aide development.

**Comment by abenari** [comment]
@domcleal added your fixes with some minor changes

**Comment by domcleal** [comment]
Merged as 5df3d51 for Foreman 1.5.0, thanks @abenari!

**Pull request closed without merging**